eval/topdown.go

```diff
@@ -0,0 +1,786 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package eval
+
+import "github.com/open-policy-agent/opa/opalog"
+import "fmt"
+
+var _ = fmt.Println
+
+// Bindings is the structure used to ground expressions.
+type Bindings map[opalog.Var]opalog.Value
+
+// TopDownContext contains the state of the evaluation process.
+//
+// TODO(tsandall): profile perf/memory usage with current approach;
+// the current approach copies the TopDownContext structure for each
+// step and binding. This avoids the need to undo steps and bindings
+// each time the proof fails but this may be too expensive.
+type TopDownContext struct {
+	Rule     *opalog.Rule
+	Bindings Bindings
+	Index    int
+	Previous *TopDownContext
+	Store    Storage
+}
+
+// Current returns the current expression to evaluate.
+func (ctx *TopDownContext) Current() *opalog.Expr {
+	return ctx.Rule.Body[ctx.Index]
+}
+
+// Step returns a new context to evaluate the next expression.
+func (ctx *TopDownContext) Step() *TopDownContext {
+	return &TopDownContext{
+		Rule:     ctx.Rule,
+		Bindings: ctx.Bindings,
+		Index:    ctx.Index + 1,
+		Previous: ctx.Previous,
+		Store:    ctx.Store,
+	}
+}
+
+// Bind returns a new TopDownContext with Bindings that map the variable to the value.
+//
+// If the caller attempts to bind a variable to itself (e.g., "x = x"), no mapping is added. This
+// is effectively a no-op.
+//
+// If the call would introduce a recursive binding, nil is returned. This represents an undefined
+// context that cannot be evaluated. E.g., "x = [x]".
+//
+// Binding a variable to some value also updates existing bindings to that variable.
+func (ctx *TopDownContext) Bind(variable opalog.Var, value opalog.Value) *TopDownContext {
+	if variable.Equal(value) {
+		return ctx
+	}
+	occurs := walkValue(value, func(other opalog.Value) bool {
+		if variable.Equal(other) {
+			return true
+		}
+		return false
+	})
+	if occurs {
+		return nil
+	}
+	cpy := *ctx
+	cpy.Bindings = make(Bindings)
+	for k, v := range ctx.Bindings {
+		cpy.Bindings[k] = plugValue(v, Bindings{variable: value})
+	}
+	cpy.Bindings[variable] = value
+	return &cpy
+}
+
+// TopDownIterator is the interface for processing contexts.
+type TopDownIterator func(*TopDownContext) error
+
+// TopDown runs the evaluation algorithm on the contxet and calls the iterator
+// foreach context that contains bindings that satisfy all of the expressions
+// inside the rule.
+func TopDown(ctx *TopDownContext, iter TopDownIterator) error {
```

Comment:
My thinking was along the lines of iterators in other languages, e.g:

```
ctx = # setup context for algorithm

# execute algorithm and process the contexts for successful proofs
for ctx in TopDown(ctx):
   # process ctx
```

In this sense, "TopDown" is like a collection and the caller iterates over it.

I'm open to alternatives, but I'm not sure reversing them would make it clearer.